h/test/realtime_test.py

```diff
@@ -0,0 +1,151 @@
+                                      exchange=exchange,
+                                      durable=False,
+                                      routing_key='annotation',
+                                      auto_delete=True)
+
+    def test_get_consumers_creates_a_consumer(self, consumer, Queue):
+        consumer_factory = mock.Mock()
+        consumer.get_consumers(consumer_factory, mock.Mock())
+        consumer_factory.assert_called_once_with(queues=[Queue.return_value],
+                                                 callbacks=[consumer.handle_message])
+
+    def test_get_consumers_returns_list_of_one_consumer(self, consumer):
+        consumer_factory = mock.Mock()
+        consumers = consumer.get_consumers(consumer_factory, mock.Mock())
+        assert consumers == [consumer_factory.return_value]
+
+    def test_handle_message_acks_message(self, consumer):
+        message = mock.Mock()
+        consumer.handle_message({}, message)
+
+        message.ack.assert_called_once_with()
+
+    def test_handle_message_calls_the_handler(self, consumer):
+        body = {'foo': 'bar'}
+        consumer.handle_message(body, mock.Mock())
+
+        consumer.handler.assert_called_once_with(body)
+
+    @pytest.fixture
+    def consumer(self):
+        return realtime.Consumer(mock.Mock(), 'annotation', mock.Mock())
+
+    @pytest.fixture
+    def Queue(self, patch):
+        return patch('h.realtime.kombu.Queue')
+
+    @pytest.fixture
+    def generate_queue_name(self, patch):
+        return patch('h.realtime.Consumer.generate_queue_name')
+
+
+class TestPublisher(object):
+    def test_publish_annotation(self, req, producer_pool):
+        payload = {'action': 'create', 'annotation': {'id': 'foobar'}}
+        producer = producer_pool['foobar'].acquire().__enter__()
+        exchange = realtime.get_exchange()
+
+        publisher = realtime.Publisher(req)
+        publisher.publish_annotation(payload)
+
+        producer.publish.assert_called_once_with(
+           payload, exchange=exchange, declare=[exchange], routing_key='annotation')
+
+    def test_publish_user(self, req, producer_pool):
+        payload = {'action': 'create', 'user': {'id': 'foobar'}}
+        producer = producer_pool['foobar'].acquire().__enter__()
+        exchange = realtime.get_exchange()
+
+        publisher = realtime.Publisher(req)
+        publisher.publish_user(payload)
+
+        producer.publish.assert_called_once_with(
+           payload, exchange=exchange, declare=[exchange], routing_key='user')
+
+    @pytest.fixture
+    def get_connection(self, patch):
+        return patch('h.realtime.get_connection')
+
+    @pytest.fixture
+    def get_exchange(self, patch):
+        return patch('h.realtime.get_exchange')
+
+    @pytest.fixture
+    def producer_pool(self, patch):
+        return patch('h.realtime.producer_pool')
+
+    @pytest.fixture
+    def req(self):
+        request = mock.MagicMock(registry=mock.MagicMock(settings={}))
```

Comment:
In circumstances like this it's probably better and easier to use a `pyramid.testing.DummyRequest` object.